Add `get_cffi_distribution_license_files()`

ames0k0 opened the pull request:
Android (termux): cffi LICENSE file is in the `licenses` folder
Fixes #187

ames0k0 committed:
Add `get_cffi_distribution_license_files()`

Android (termux): cffi LICENSE file is in the `licenses` folder

ames0k0 commented:
Ok. I can't import it
```
Python 3.12.8 (main, Dec  4 2024, 22:36:35) [Clang 18.0.3 (https://android.googlesource.com/toolchain/llvm-project d8003a456 on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import coincurve
Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
    File "/data/data/com.termux/files/home/trash/env/lib/python3.12/site-packages/coincurve/__init__.py", line 1, in <module>
        from coincurve.context import GLOBAL_CONTEXT, Context
    File "/data/data/com.termux/files/home/trash/env/lib/python3.12/site-packages/coincurve/context.py", line 6, in <module>
        from coincurve._libsecp256k1 import ffi, lib
ImportError: dlopen failed: cannot locate symbol "_Py_NoneStruct" referenced by "/data/data/com.termux/files/home/trash/env/lib/python3.12/site-packages/coincurve/_libsecp256k1.cpython-312.so"...
>>>
```

robertkirkman commented:
`ImportError: dlopen failed: cannot locate symbol "_Py_NoneStruct"`

To fix this error, please use:

```bash
LDFLAGS=-lpython3.12 pip install --force-reinstall coincurve==20.0.0
```

When Termux updates to Python 3.13, it will be required to rerun the same command again, but with `-lpython3.13` instead.

MementoRC commented:
@DimitriPapadopoulos Your latest push to master seemed to have failed all checks. This PR should be merged (we included it in the conda-forge package to pass the python 3.14 support)

MementoRC commented:
@DimitriPapadopoulos My mistake, never mind.

ofek commented:
I think this might pass after small changes are made to satisfy the static analysis.

MementoRC committed:
(fix) staticmethod

MementoRC committed:
(fix) Improve license file identification process

Refactor license file detection logic in hatch_build.py

DimitriPapadopoulos commented:
Looks good, now one last issue to fix on macOS:
```
   delocate.libsana.DelocationError: Library dependencies do not satisfy target MacOS version 10.9:
  /private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/tmpu9qpg5w_/wheel/coincurve/_cffi_backend.cpython-39-darwin.so has a minimum target of 10.13
  Set the environment variable 'MACOSX_DEPLOYMENT_TARGET=10.13' to update minimum supported macOS for this wheel.
Error: Command delocate-wheel --require-archs x86_64 -w /private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/cibw-run-zknnvqsc/cp39-macosx_x86_64/repaired_wheel -v /private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/cibw-run-zknnvqsc/cp39-macosx_x86_64/built_wheel/coincurve-21.0.0-cp39-cp39-macosx_10_9_x86_64.whl failed with code 1. 
```

DimitriPapadopoulos commented on hatch_build.py:
```diff
@@ -22,13 +22,32 @@ class CustomBuildHook(BuildHookInterface):
+            if f.name != "LICENSE":
+                continue
+
+            current = f.parent
+            while current and not current.name.endswith(".dist-info"):
+                current = current.parent
+                if current is None or len(current.parts) == 0:
+                    break
```
Not entirely equivalent, but doesn't it do the job?
```suggestion
            current = f.parent
            while current and current.parts and not current.name.endswith(".dist-info"):
                current = current.parent
```

DimitriPapadopoulos commented on hatch_build.py:
```diff
@@ -22,13 +22,32 @@ class CustomBuildHook(BuildHookInterface):
 
+    @staticmethod
+    def get_cffi_distribution_license_files() -> list[PackagePath]:
+        license_files = []
+
+        dist_files = distribution("cffi").files or []
+        for f in dist_files:
+            if f.name != "LICENSE":
+                continue
```
```suggestion
        license_files = [
            file
            for file in distribution("cffi").files
            if file.name == "LICENSE"
        ]
```

DimitriPapadopoulos reviewed (commented):
A few suggestions to simplify code.

DimitriPapadopoulos commented on hatch_build.py:
```diff
@@ -22,13 +22,32 @@ class CustomBuildHook(BuildHookInterface):
+            if f.name != "LICENSE":
+                continue
+
+            current = f.parent
+            while current and not current.name.endswith(".dist-info"):
+                current = current.parent
+                if current is None or len(current.parts) == 0:
+                    break
```
Not entirely equivalent, but doesn't it do the job?
```suggestion
            current = f.parent
            while current and current.parts and not current.name.endswith(".dist-info"):
                current = current.parent
```

ames0k0 replied:
`current` is always "truthly" (`PackagePath('.')`), so it can be removed

```python
while current.parts and not current.name.endswith(".dist-info"):
```

MementoRC commented:
@ames0k0 Please, can you apply the macos-14 update to this PR?

ames0k0 committed:
github actions: Updated jobs `macos-wheels-x86_64`

Set the environment variable 'MACOSX_DEPLOYMENT_TARGET=10.13'
to update minimum supported macOS for this wheel.

ames0k0 committed:
Merge branch 'ofek:master' into issues/187

ofek commented:
```
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/y6/nj790rtn62lfktb1sh__79hc0000gn/T/cibw-run-uqf9ogxo/cp39-macosx_x86_64/venv-test-x86_64/lib/python3.9/site-packages/coincurve/__init__.py", line 1, in <module>
      from coincurve.context import GLOBAL_CONTEXT, Context
    File "/private/var/folders/y6/nj790rtn62lfktb1sh__79hc0000gn/T/cibw-run-uqf9ogxo/cp39-macosx_x86_64/venv-test-x86_64/lib/python3.9/site-packages/coincurve/context.py", line 6, in <module>
      from coincurve._libsecp256k1 import ffi, lib
  ImportError: dlopen(/private/var/folders/y6/nj790rtn62lfktb1sh__79hc0000gn/T/cibw-run-uqf9ogxo/cp39-macosx_x86_64/venv-test-x86_64/lib/python3.9/site-packages/coincurve/_cffi_backend.cpython-39-darwin.so, 0x0002): tried: '/private/var/folders/y6/nj790rtn62lfktb1sh__79hc0000gn/T/cibw-run-uqf9ogxo/cp39-macosx_x86_64/venv-test-x86_64/lib/python3.9/site-packages/coincurve/_cffi_backend.cpython-39-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/private/var/folders/y6/nj790rtn62lfktb1sh__79hc0000gn/T/cibw-run-uqf9ogxo/cp39-macosx_x86_64/venv-test-x86_64/lib/python3.9/site-packages/coincurve/_cffi_backend.cpython-39-darwin.so' (no such file), '/private/var/folders/y6/nj790rtn62lfktb1sh__79hc0000gn/T/cibw-run-uqf9ogxo/cp39-macosx_x86_64/venv-test-x86_64/lib/python3.9/site-packages/coincurve/_cffi_backend.cpython-39-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))
```

ames0k0 committed:
github actions: Updated workflows

build.yml
	- Removed env variable `MACOSX_DEPLOYMENT_TARGET`
verify_shared_build.yml
	- Updated MacOS version from `13` to `14`

MementoRC commented:
My mistake, I misread the Github notification:
```
For users that require the x86_64 (Intel) architecture, we have added a new label to support standard runner users. Jobs can be migrated to one of the following labels:
– macos-15-intel (new)
– macos-14-large
– macos-latest-large or macos-15-large
```
the CMake must be building a `arm64` _cffi_backend.so on the arm runner (although it references x86_64 in the logs - odd)
Can we try `macos-14-large`?

ofek commented:
I'm so sorry that I don't have time to look into this myself but could someone please briefly summarize the macOS issue?

MementoRC commented:
Ok, I don't understand the error at all, somehow python arm64 is installed by the setup-python@v5 on the macos-14:
`##[debug]Set output python-path = /Users/runner/hostedtoolcache/Python/3.12.10/arm64/bin/python` , this causes the _cffi_backend.so to be an arm64 binary and cannot be loaded when we try to relocate the wheel

MementoRC commented:
> `macos-14-large` is only for paying customers. `macos-14` worked for me.

What do you mean? in your fork this PR passes? Is it a random issue due to some runners (I tried to rerun by same result)

MementoRC commented:
This is odd:
```
Runner Image
  Image: macos-14-arm64
  Version: 20250928.1833
  Included Software: https://github.com/actions/runner-images/blob/macos-14-arm64/20250928.1833/images/macos/macos-14-arm64-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/macos-14-arm64%2F20250928.1833
```

MementoRC commented:
There is this:
```
Workflows that depend on x86_64 (Intel CPU) can be migrated to use the macos-15-intel label, which is currently planned as the last supported image based on Intel CPU.
```

MementoRC commented:
There is this:
```
Workflows that depend on x86_64 (Intel CPU) can be migrated to use the macos-15-intel label, which is currently planned as the last supported image based on Intel CPU.
```

MementoRC commented:
> * **`macos-13`** has an issue with target minimal mac0S

Then we should either try `macos-13` with the `MACOSX_DEPLOYEMENT_TARGET: 10.13` env var or `macos-15-intel`

DimitriPapadopoulos commented:
> > `macos-14-large` is only for paying customers. `macos-14` worked for me.
> 
> What do you mean? in your fork this PR passes? Is it a random issue due to some runners (I tried to rerun by same result)

Whenever I tried **`macos-14-large`** in open source GitHub repositories, it failed. I had to switch to **`macos-14`**.

MementoRC committed:
Change macOS runner version from 14 to 13

MementoRC committed:
Merge branch 'master' into issues/187

DimitriPapadopoulos commented:
As far I know, macOS build tools can target lower minimal macOS versions. This is handled through different build options or environment variables (such as `MACOSX_DEPLOYEMENT_TARGET`).

MementoRC committed:
Fix typo in MACOSX_DEPLOYMENT_TARGET variable

MementoRC committed:
Update build.yml

DimitriPapadopoulos commented:
Interesting, after bumping the target minimal macOS version from 10.9 (the default on `macos-13` as far as I can understand) to 10.13 using `MACOSX_DEPLOYEMENT_TARGET`, to match the target minimal macOS version of the CFFI backend, we need with the same error on `macos-13` as in `macos-14`:
```
site-packages/coincurve/_cffi_backend.cpython-39-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))
```

We need a way to obtain the CFFI backend for the x86_64 architecture!

MementoRC commented:
There was a typo in the variable name. I concurrently updated to `macos-15-intel` on `build.yml` and corrected the variable in the shared.yml, hopefully one will work.
I think that `macos-14` is an arm64 runner, so CMake builds for that target with the python binding to arm64, but somehow, cibuildwheel can build an x86_64 and does until it needs to load the _cffi_backend.so and fails

ofek commented on .github/workflows/build.yml:
```suggestion
        CIBW_ARCHS_MACOS: x86_64
        MACOSX_DEPLOYMENT_TARGET: 10.13
```

ofek committed:
Apply suggestion from @ofek

DimitriPapadopoulos commented on .github/workflows/verify_shared_build.yml:
Perhaps bump to `macos-15` (and `macos-15-intel` where needed) for consistency and to use the most recent build tools.

DimitriPapadopoulos commented on hatch_build.py:
```diff
@@ -22,13 +22,32 @@ class CustomBuildHook(BuildHookInterface):
+            while current and not current.name.endswith(".dist-info"):
+                current = current.parent
+                if current is None or len(current.parts) == 0:
+                    break
```
Can `current.parent` be `None`? Can't find any mention of `None` in https://docs.python.org/fr/3.13/library/pathlib.html#pathlib.PurePath.parent.

ofek replied:
I would have to test, concrete paths can.

ofek commented on hatch_build.py:
```diff
@@ -22,13 +22,32 @@ class CustomBuildHook(BuildHookInterface):
+
+        dist_files = distribution("cffi").files or []
+        for f in dist_files:
+            if f.name != "LICENSE":
+                continue
+
+            current = f.parent
+            while current and not current.name.endswith(".dist-info"):
+                current = current.parent
+                if current is None or len(current.parts) == 0:
+                    break
+
+            if current and current.name.endswith(".dist-info"):
+                license_files.append(f)
```
```suggestion
            if f.name == "LICENSE" and parts[0].endswith(".dist-info"):
                license_files.append(f)
                break
```

ofek commented on .github/workflows/build.yml:
```diff
@@ -140,6 +140,7 @@ jobs:
       uses: pypa/cibuildwheel@v2.23
```
```suggestion
      uses: pypa/cibuildwheel@v3.2.0
```

ofek commented on .github/workflows/verify_shared_build.yml:
```diff
@@ -96,6 +96,7 @@ jobs:
       uses: pypa/cibuildwheel@v2.23
```
```suggestion
      uses: pypa/cibuildwheel@v3.2.0
```

DimitriPapadopoulos commented on .github/workflows/build.yml:
```diff
@@ -140,6 +140,7 @@ jobs:
       uses: pypa/cibuildwheel@v2.23
       env:
         CIBW_ARCHS_MACOS: x86_64
+        MACOSX_DEPLOYMENT_TARGET: 10.13
```
Let's leave it for now. We can try again after bumping all GitHub Actions, and once CI is back on its tracks.

ofek commented on .github/workflows/verify_shared_build.yml:
```suggestion
    runs-on: macos-15-intel
```

ofek committed:
Update hatch_build.py

ofek committed:
Update .github/workflows/build.yml

ofek committed:
Update .github/workflows/verify_shared_build.yml

ofek committed:
Apply suggestion from @ofek

ofek committed:
Apply suggestion from @ofek

ofek commented:
There's another error, both are failing the same way now.

DimitriPapadopoulos commented:
About the last error: cbuildwheels 3.2.0, pulls Python 3.14 (free-threaded):
```
  /Library/Frameworks/PythonT.framework/Versions/3.14/include/python3.14t/Python.h:51:4: error: "The limited API is not currently supported in the free-threaded build"
     51 | #  error "The limited API is not currently supported in the free-threaded build"
        |    ^
  In file included from /private/var/folders/5k/fq5ptf2906bgfn1krb78bmj00000gn/T/tmpzcln63g5/build/_gen_c_file/secp256k1_cffi_bindings.c:58:
  In file included from /Library/Frameworks/PythonT.framework/Versions/3.14/include/python3.14t/Python.h:81:
  /Library/Frameworks/PythonT.framework/Versions/3.14/include/python3.14t/object.h:158:5: error: unknown type name 'PyMutex'
    158 |     PyMutex ob_mutex;           // per-object lock
        |     ^
```
Is there a way to downgrade for now?

MementoRC commented:
we could use:
```
# Disable free-threading builds on all platforms
skip = "cp3??t-*"
```
the non-threaded 3.14 passes

MementoRC committed:
Skip building specific CPython wheels on macOS

MementoRC committed:
Update verify_shared_build.yml

MementoRC committed:
Merge branch 'master' into issues/187

MementoRC committed:
Add Python 3.14 to build matrix

MementoRC committed:
Remove Python version 3.14 from CI workflow

MementoRC commented:
3.14 does not seem fully supported yet but in rc-3

ofek reviewed (approved):
Great job everybody, thanks!

ttodua commented:
Hi @ofek ,
when do you think to push a new release?

prusnak commented:
+1 for new release with a fix

darosior commented:
Hi @ofek could you publish a release with this fix? This is a blocker for downstream projects.

mbrousset-ledger commented:
Hello @ofek ! Just wondering if this repository is still maintained as a few other repositories are blocked with this ?
